docs: ringing docs improvements (#505)
GetStream/stream-video-flutter: docusaurus/docs/Flutter/05-advanced/02-ringing.mdx

You now need to upload your Firebase credentials using the Stream dashboard.

- Go to the dashboard of your video project at the [Stream website](https://dashboard.getstream.io).

- Open the **Push Notifications** tab under **Video & Audio**.

- Select **New Configuration** and select **Firebase**.

![Firebase Configuration](../assets/advanced_assets/firebase_config.png)

> - Add a name for your push provider in the **Name** field. You will use this name later in the code to identify which provider to use for android notifications.

Reviewer: ```suggestion
- Add a name for your push provider in the **Name** field. You will use this name later in the code to identify which provider to use for Android notifications.
```

- Add your previously generated Firebase Credentials in the **Credentials JSON** field.

> - Enable this provider using toggle button

Reviewer: ```suggestion
- Enable this provider using toggle button.
```

- Click **Create** and your push provider should be ready.

#### Step 3 - Add dependencies to your app

> To integrate push notifications in your Flutter app, you need to use the **firebase_messaging** package.

Author: Good question :) I think I wanted to use it somewhere below but didn't. I'm putting it back in

To handle calls from a terminated state, we need to add some additional code.

In a high-level widget, add this method and call it from the `initState()` method:
- add navigator key to MaterialApp widget: `navigatorKey: _navigatorKey`

```dart
final _navigatorKey = GlobalKey<NavigatorState>();

@override
void initState() {
  //...
  _tryConsumingIncomingCallFromTerminatedState();
}

void _tryConsumingIncomingCallFromTerminatedState() {
  if (_navigatorKey.currentContext == null) {
    // App is not running yet. Postpone consuming after app is in the foreground
    WidgetsBinding.instance.addPostFrameCallback((timeStamp) {
      _consumeIncomingCall();
    });
  } else {
    // no-op. If the app is already running we'll handle it via events
  }
}

Future<void> _consumeIncomingCall() async {
  final calls = await StreamVideo.instance.pushNotificationManager?.activeCalls();
  if (calls == null || calls.isEmpty) return;

  final call = await StreamVideo.instance.consumeIncomingCall(
    uuid: calls.first.uuid,
    cid: calls.first.callCid,
  );
}
```

#### Step 6 - Request notification permission from user

> For Android 13+ you need to request post notification permission. You can do it using [permission_handler](https://pub.dev/packages/permission_handler) package.

Reviewer: ```suggestion
For Android 13+ you need to request the `POST_NOTIFICATIONS` permission. You can do it using the [permission_handler](https://pub.dev/packages/permission_handler) package.
```

- Open the **Push Notifications** tab under **Video & Audio**.

- Select **New Configuration** and select **APN**.

![APNs Configuration](../assets/advanced_assets/apns_config.png)

- Add a name for your push provider in the **Name** field. This is the name used while setting up iOS push notifications in the code we did [earlier](#Step-5---Add-code-to-listen-to-push-notifications)

- Add your previously generated P12 file with your additional Apple information.

- Enable this provider using toggle button

- Click **Create** and your push provider should be ready.

#### Step 3 - Add dependencies

There are no dependencies on the Flutter side that you need to add specifically for iOS.

#### Step 4 - Add native permissions

Add these permissions to `Info.plist` in order to support video calling:

```plist
<key>NSCameraUsageDescription</key>
<string>$(PRODUCT_NAME) Camera Usage!</string>
<key>NSMicrophoneUsageDescription</key>
<string>$(PRODUCT_NAME) Microphone Usage!</string>
<key>UIApplicationSupportsIndirectInputEvents</key>
<true/>
<key>BGTaskSchedulerPermittedIdentifiers</key>
<array>

>   <string>$(PRODUCT_BUNDLE_IDENTIFIER)</string>

Reviewer: ```suggestion
	 <string>$(PRODUCT_BUNDLE_IDENTIFIER)</string>
```

### Integrating APNs for iOS

#### Step 1 - Get the iOS certificate for push notifications

- Generate push notification service key [here](https://developer.apple.com/account/resources/certificates/add). Make sure you select **Apple Push Notifications service SSL (Sandbox & Production)**.

- You will need to create **Certificate Signing Request** - [follow this steps](https://developer.apple.com/help/account/create-certificates/create-a-certificate-signing-request)

- Convert the *aps.cer* file you created in the last step to a .p12 certificate file using keychain access. Make sure that you configure no password for the p12 file.
  - Add app.cer to login keychain
  - Find it in Certificate tab, right click and export as .p12 file
  - Remember not to set any password while exporting

#### Step 2 - Upload the certificate and create a push provider

- Go to the dashboard of your video project at the [Stream website](https://dashboard.getstream.io).

- Open the **Push Notifications** tab under **Video & Audio**.

- Select **New Configuration** and select **APN**.

![APNs Configuration](../assets/advanced_assets/apns_config.png)

> - Add a name for your push provider in the **Name** field. This is the name used while setting up ios push notifications in the code we did [earlier](#Step-5---Add-code-to-listen-to-push-notifications)

Reviewer: ```suggestion
- Add a name for your push provider in the **Name** field. This is the name used while setting up iOS push notifications in the code we did [earlier](#Step-5---Add-code-to-listen-to-push-notifications)
```

#### Step 4 - Add native permissions

Add these permissions to `Info.plist` in order to support video calling:

```plist
<key>NSCameraUsageDescription</key>
<string>$(PRODUCT_NAME) Camera Usage!</string>
<key>NSMicrophoneUsageDescription</key>
<string>$(PRODUCT_NAME) Microphone Usage!</string>
<key>UIApplicationSupportsIndirectInputEvents</key>
<true/>
<key>BGTaskSchedulerPermittedIdentifiers</key>
<array>
  <string>$(PRODUCT_BUNDLE_IDENTIFIER)</string>
</array>
<key>UIBackgroundModes</key>
<array>
	<string>audio</string>
	<string>fetch</string>
	<string>processing</string>
	<string>remote-notification</string>
	<string>voip</string>
</array>
```

#### Step 5 - Add code to handle CallKit events

In a high-level widget in your app, add this code to listen to FCM messages:

```
@override
void initState() {
  ...
  _observeFcmMessages()
}

_observeFcmMessages() {
  FirebaseMessaging.onBackgroundMessage(_firebaseMessagingBackgroundHandler);
  _fcmSubscription = FirebaseMessaging.onMessage.listen(_handleRemoteMessage);
}
```

In a high-level widget in your app, add this code to listen to CallKit events:

> - remember to implement navigation in a marked line

Reviewer: This line is a bit unclear. Can we rephrase or add it later as a note?
